### PR TITLE
STAT-64 Chain to mongo 172

### PR DESCRIPTION
**Pull request by heifner**
**STAT-64 - Current implementation:**

* chain_controller
  * Added {{signal<void(const signed_block&)> applied_irreversible_block}}

* db_plugin
  * Adds to mongodb
    * Schema: https://docs.google.com/document/d/1E6BQVX1E8jsCzJIpFWV_B3hfA7Br6iiFlUiWZJpJ3zM/edit?ts=59d29be1#heading=h.sgz9di1jhkvk
    * Block, Transaction, Message
    * Account eos_balance, abi

**Future implementation:**

* eosd running in delay mode so that it only processes irreversible blocks
  * Options:
    1. eosd setup so that it only pushes block from P2P to chain_controller when the block is irreversible
       * eosd would need to cache blocks until the irreversible block num indicates they can be pushed to chain_controller
       * eosd would be in a state of constant "syncing"
       * chainbase would be set up to push to mongodb
       * Simpler configuration since only one special eosd required
    2. a trusted eosd setup so that it only broadcasts irreversible blocks
       * delayed eosd connected to trusted eosd would then simply run as normal
       * the delayed eosd would be set up to push to mongodb

* Add hook to chainbase so that commit pushes into mongodb
  * Either a hook given to chainbase OR a replacement of chainbase

### CODE REVIEW

File: libraries/chain/chain_controller.cpp [hunk 1]
```diff
@@ -1224,15 +1227,25 @@ void chain_controller::initialize_chain(chain_initializer_interface& starter)
          for (int i = 0; i < 0x10000; i++)
             _db.create<block_summary_object>([&](block_summary_object&) {});
 
+         signed_block block{};
```
Comment (brianjohnson5972) [inline libraries/chain/chain_controller.cpp]: Please add a comment to indicate what you are doing here.

File: libraries/chain/chain_controller.cpp [hunk 2]
```diff
@@ -1273,6 +1290,12 @@ chain_controller::~chain_controller() {
 void chain_controller::replay() {
    ilog("Replaying blockchain");
    auto start = fc::time_point::now();
+
+   auto on_exit = fc::make_scoped_exit([this](){
```
Comment (brianjohnson5972) [inline libraries/chain/chain_controller.cpp]: Limit lambda capturing to just one needed parameter.

Comment (heifner) [reply libraries/chain/chain_controller.cpp]: Done.

File: libraries/chain/chain_controller.cpp (hunk 1 above)
Comment (heifner) [reply libraries/chain/chain_controller.cpp]: Added comment:
`         // create a dummy block and cycle for our dummy transactions to send to applied_irreversible_block below
`

**Pull request merged**